### ISSUE TITLE
Repair doc render regression

### ISSUE THREAD

**Issue by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#3476]
When we updated our builder to generate documentation with Swift 6 by default, we accidentally also picked up a performance regression with the included JS: https://github.com/swiftlang/swift-docc-render/issues/905

Since then we've been generating doc archives than can lead to bad render performance in some cases (doc archives with many symbols on their root page).

While a fix is available on nightly toolchains, we can't easily adopt those for documentation generation.

We only really have two options:
- roll back to generating documentation with Swift 5.10 by default
- wait for the next Xcode release with the fix (probably Xcode 16.2 but to be confirmed)

We are currently (Nov 5) referencing 871 doc archives with the problematic JS.

Query to identify builds/versions:

```sql
select builder_version, platform, swift_version, b.status, d.status, v.latest
from builds b
join doc_uploads d on d.build_id = b.id
join versions v on b.version_id = v.id
where builder_version like '4.55.%' or builder_version like '4.56.%'
and d.status = 'ok'
```

**Comment by finestructure** [comment]
cc @franklinsch

**Comment by heckj** [comment]
I wonder if the Xcode builds that generate documentation can be "overridden" for what DOCC_RENDERER content they use with the environment variable setup:
```bash
DOCC_HTML_DIR=/path/to/swift-docc-render-artifact/dist
```

It would be quite the annoying shim in the builders, but may work around the issue before Xcode 6.2 is available.

**Comment by heckj** [comment]
I learned about it from part of the code that supports testing/debugging, and there was a mention of it as a workaround by mporitz in Miguel's bug report. The gist, as I understand it, is that if that environment variable is set and passed into the build, docc will use it to determine where to pull the javascript files from. I haven't tried it with an `xcodebuild` command, but I expect it will likely do the job. There's a reference to this whole set up about half-way down in https://github.com/swiftlang/swift-docc/blob/main/CONTRIBUTING.md

**Comment by mportiz08** [comment]
> That sounds like it's worth a try. Is there some documentation somewhere how that works exactly? Xcode 16.2 seems fairly close but this might be the way to go if the fix doesn't make it in there.

Yep, it's here: https://github.com/swiftlang/swift-docc/blob/main/CONTRIBUTING.md#2-set-the-path-to-your-renderer

I don't know how much of a disruption it might be, but you could potentially clone the latest [swift-docc-artifact][1] which has a prebuilt version of the renderer that gets built with CI using the main branch of swift-docc-render, and setting `DOCC_HTML_DIR=/path/to/swift-docc-render-artifact/dist` would allow you to use the latest version of DocC-Render to override the copy included with the toolchain you're using.

With that being said, I recently cherry-picked the fix for the regression to a release/6.0 branch so that the fix will be available on the next possible toolchain. Apologies for the troubles!

[1]: https://github.com/swiftlang/swift-docc-render-artifact

**Comment by finestructure** [comment]
Thanks a lot for the additional info, @heckj and @mportiz08 . When you say "available on the next possible toolchain" does that mean it would be in Xcode 16.2?

Regardless, I'm going to try out your suggested fix. It doesn't sound like it'd be too complicated to integrate 🙂 Thanks again for the details!

**Comment by finestructure** [comment]
Just to clarify, @heckj , you [said on Slack](https://swift-open-source.slack.com/archives/C04PCMXMBD0/p1730737231567839):

> I think this file might be a good marker file: [104.fe5974d0.js](https://github.com/swiftlang/swift-docc-render-artifact/blob/release/6.0.2/dist/js/104.fe5974d0.js) (from https://github.com/swiftlang/swift-docc-render-artifact/tree/release/6.0.2/dist/js)

When I diff the current doc gen archive's `js` dir with `dist/js` from `swift-docc-render-artifact`, this file is present in both:

<img width="1862" alt="Screenshot 2024-11-08 at 08 24 14" src="https://github.com/user-attachments/assets/83f389b9-a8dd-456c-abd8-1b88d8feea4b">

Seems to me I should rather be looking for one of `842.*`, `866.*` `documentation-topic.*`, `index.*`, `topic.*`, `tutorials-overview.*` to change.

This is for the purpose of writing a test to ensure a fixed archive is being generated.

**Comment by heckj** [comment]
You've got a better link on the right files with the diff. I only scanned the repository files through github looking for recently changed files, and noted that the one that starts with `104.` was updated, so I took that as a possible marker. 

I wasn't sure how those file are generated, and which file names are useful to use as markers, assuming that they _do_ change names when the contents used to generate them update. Based on what you've got above, the ones that start with `842.`, `866.`, or `documentation-topics.` look like the ideal candidates. So yeah - I totally agree with you there!

**Comment by finestructure** [comment]
I've deployed a new builder version 4.57.0 that pulls in the artefacts from `swift-docc-render-artifact`.

In the process of regenerating the ~1870 doc builds with the faulty JS. I'll close this issue when the reprocessing is complete.

Thanks again @heckj  and @mportiz08 for your help!

**Comment by finestructure** [comment]
All affected doc archives have been regenerated.
